docs/backend-system/building-backends/08-migrating.md

```diff
@@ -7,3 +7,238 @@ description: How to migrate existing backends to the new backend system
+way that plugins and their dependencies are wired up, leading to a significantly
+simplified backend package that rarely if ever needs to change when plugins or
+their dependencies evolve. You generally don't have to convert all of your
+internal plugins and support classes themselves to the backend system first -
+the migration here will mostly deal with wiring and using compatibility wrappers
+where possible in the backend package itself. We hope that you will find that
+you end up with a much smaller, easier to understand, and easier to maintain
+package as a result of these steps, and then being able to [migrate
+plugins](../building-plugins-and-modules/08-migrating.md) as a separate
+endeavour later.
+
+# Overall Structure
+
+Your typical backend package has a few overall component parts:
+
+- An `index.ts` file that houses all of the creation and wiring together of all
+  of the plugins and their dependencies
+- A `types.ts` file that defines the "environment", i.e. the various
+  dependencies that get created by the backend and passed down into each plugin
+- A `plugins` folder which has one file for each plugin, e.g.
+  `plugins/catalog.ts`
+
+The index file has this overall shape:
+
+```ts
+import todo from './plugins/todo'; // repeated for N plugins
+
+function makeCreateEnv(config: Config) {
+  return (plugin: string): PluginEnvironment => {
+    // ... build per-plugin environment
+  };
+}
+
+async function main() {
+  // ... early init
+  const createEnv = makeCreateEnv(config);
+  const todoEnv = useHotMemoize(module, () => createEnv('todo')); // repeated for N plugins
+  const apiRouter = Router();
+  apiRouter.use('/todo', await todo(todoEnv)); // repeated for N plugins
+  // ... return composite router
+}
+
+module.hot?.accept();
+main().catch(...);
+```
+
+# Migrating the Index File
+
+This migration will try to leave the `plugins` folder unchanged initially, first
+focusing on removing the environment type and reducing the index file to its
+bare minimum. Then as a later step, we can reduce the `plugins` folder bit by
+bit, replacing those files generally with one-liners in the index file instead.
+
+Let's start by establishing the basis of your new index file. You may want to
+comment out its old contents, or renaming the old file to `index.backup.ts` for
+reference and making a new blank one to work on - whichever works best for you.
+These are our new blank contents in the index file:
+
+```ts
+// packages/backend/src/index.ts
+import { createBackend } from '@backstage/backend-defaults';
+
+const backend = createBackend();
+backend.start();
+```
+
+Note that the environment builder and the `main` dance are entirely gone.
+
+We'll also want to add some backend system packages as dependencies. Run the
+following command:
+
+```bash
+# from the repository root
+yarn add --cwd packages/backend @backstage/backend-common @backstage/backend-defaults @backstage/backend-plugin-api
```

Comment:
```suggestion
yarn add --cwd packages/backend @backstage/backend-defaults @backstage/backend-plugin-api
```

`backend-common` should always be present already

Reply:
yeah i had it for completeness (it's used in the examples) but you are right that it's better to leave out